### PR TITLE
Flag values that are truncated in qstat

### PR DESCRIPTION
**Pull request by vstumpf**
#### Describe Bug or Feature
Qstat truncates values if it is longer than the column width, leading to some confusion.

#### Describe Your Change
If qstat will truncate a value, substitute the last character with an asterisk `*`.

#### Link to Design Doc
https://pbspro.atlassian.net/wiki/spaces/PD/pages/1289420812/Flag+fields+where+data+is+truncated+in+qstat

#### Attach Test and Valgrind Logs/Output
[after-fix.txt](https://github.com/PBSPro/pbspro/files/3242478/after-fix.txt)

### CODE REVIEW

File: src/cmds/qstat.c [hunk 1]
```diff
@@ -302,6 +304,7 @@ states(char *string, char *q, char *r, char *h, char *w, char *t, char *e, int l
 			*c = '\0';
 			if (strlen(s) > (size_t)len) {
 				f = s + len;
+				*(f - 1) = DISPLAY_TRUNC_FLAG;
```
Comment (mike0042) [inline src/cmds/qstat.c]: Perhaps a bit pedantic, but please check (f > s) before you assign *(f - 1).

File: src/cmds/qstat.c [hunk 2]
```diff
@@ -545,6 +548,32 @@ prt_attr(char *name, char *resource, char *value, int one_line) {
...
+ *
+ * @param[in] value - value that may be truncated
+ * @param[in] len - non-wide length that must be met
+ * @param[in] wide_len - wide length that must be met
+ * @param[in] wide - whether or not wide format is used
+ *
+ */
+
+static void
+trunc_value(char *value, int len, int wide_len, int wide)
```
Comment (mike0042) [inline src/cmds/qstat.c]: C99 allows you to use bool for the wide parameter. Why do you need two length parameters if you'll only ever use one or the other? Can we remove wide_len?

File: src/cmds/qstat.c [hunk 3]
```diff
@@ -545,6 +548,32 @@ prt_attr(char *name, char *resource, char *value, int one_line) {
...
+{
+	if (wide) {
+		if (strlen(value) > wide_len) {
+			*(value + wide_len - 1) = DISPLAY_TRUNC_FLAG;
+		}
+	} else {
+		if (strlen(value) > len) {
+			*(value + len - 1) = DISPLAY_TRUNC_FLAG;
```
Comment (mike0042) [inline src/cmds/qstat.c]: You should ensure (len > 0) before you make this assignment.

File: src/cmds/qstat.c [hunk 4]
```diff
@@ -924,10 +961,14 @@ altdsp_statjob(struct batch_status *pstat, struct batch_status *prtheader, int a
 		else
 			timeval = usecput ? eltimecpu : eltimewal;
 
-
 		if (wide) {
+			if (strlen(pstat->name) > SIZEJOBID) {
```
Comment (mike0042) [inline src/cmds/qstat.c]: Should this be jname instead of pstat->name?

File: src/cmds/qstat.c [hunk 5]
```diff
@@ -942,11 +983,21 @@ altdsp_statjob(struct batch_status *pstat, struct batch_status *prtheader, int a
 				timeval);
 		} else {
 			if(how_opt & ALT_DISPLAY_INCR_WIDTH) {
```
Comment (mike0042) [inline src/cmds/qstat.c]: Please add a space after the if.

File: src/cmds/qstat.c [hunk 6]
```diff
@@ -942,11 +983,21 @@ altdsp_statjob(struct batch_status *pstat, struct batch_status *prtheader, int a
...
+				if (strlen(pstat->name) > 20) {
```
Comment (mike0042) [inline src/cmds/qstat.c]: Should this be jname instead of pstat->name?

File: src/cmds/qstat.c [hunk 7]
```diff
@@ -1362,16 +1413,19 @@ display_statjob(struct batch_status *status, struct batch_status *prtheader, int
 				if (wide) {
 					if (l > SIZEJOBID) {
 						c = p->name + SIZEJOBID;
+						*(c - 1) = DISPLAY_TRUNC_FLAG;
```
Comment (mike0042) [inline src/cmds/qstat.c]: This should be safe since we know SIZEJOBID > 0.

File: src/cmds/qstat.c [hunk 8]
```diff
@@ -1362,16 +1413,19 @@ display_statjob(struct batch_status *status, struct batch_status *prtheader, int
...
 						*c = '\0';
 					}
 				} else if (how_opt & ALT_DISPLAY_INCR_WIDTH) {
 					if (l > (PBS_MAXSEQNUM+10)) {
 						c = p->name + PBS_MAXSEQNUM + 10;
+						*(c - 1) = DISPLAY_TRUNC_FLAG;
 						*c = '\0';
 					}
 				} else {
 					if (l > (PBS_MAXSEQNUM+5)) {
```
Comment (mike0042) [inline src/cmds/qstat.c]: Please add space around the plus sign.

File: src/cmds/qstat.c (hunk 4 above)
Comment (vstumpf) [reply src/cmds/qstat.c]: jname hadn't been filled out yet.

File: src/cmds/qstat.c (hunk 2 above)
Comment (vstumpf) [reply src/cmds/qstat.c]: If I change `trunc_value` to only take one length, then everywhere I call it I have to put `if (wide) ... else` blocks.
I kept wide as an int to follow the style of the file, but I can change it if you want.

File: src/cmds/qstat.c (hunk 6 above)
Comment (vstumpf) [reply src/cmds/qstat.c]: Same as above, jname hasn't been filled yet.

File: src/cmds/qstat.c (hunk 2 above)
Comment (mike0042) [reply src/cmds/qstat.c]: Ahh... thanks for explaining, that makes sense.

File: src/cmds/qstat.c [hunk 9]
```diff
@@ -836,25 +867,32 @@ altdsp_statjob(struct batch_status *pstat, struct batch_status *prtheader, int a
 		while (pat) {
 			if (strcmp(pat->name, ATTR_N) == 0) {
 				jobn = pat->value;
+				trunc_value(jobn, 10, SIZEJOBNAME, wide);
```
Comment (mike0042) [inline src/cmds/qstat.c]: We should define all field widths as macros, as we do for the wide values.

File: src/cmds/qstat.c [hunk 10]
```diff
@@ -102,6 +102,8 @@ char *cnvt_est_start_time(char *start_time, int shortform);
 #define HPCBP_EXEC_TAG  "jsdl-hpcpa:Executable"
 #define CHAR_LINE_LIMIT 78 /* maximum number of characters which can be printed on a line */
 
+#define DISPLAY_TRUNC_FLAG '*'
```
Comment (bhroam) [inline src/cmds/qstat.c]: This really isn't a flag.  A flag changes how something works.  This is just a character that indicates something.  Maybe DISPLAY_TRUNC_CHAR?

File: src/cmds/qstat.c [hunk 11]
```diff
@@ -545,6 +549,32 @@ prt_attr(char *name, char *resource, char *value, int one_line) {
 #define SIZETSK 5	/* length of tsk in printf */
 #define SIZEUSER 15	/* length of user name in printf */
 
+
+/**
+ * @brief
+ *	Check if value is too long, truncate and append with * if needed
```
Comment (bhroam) [inline src/cmds/qstat.c]: Don't say '*'.  It's dependent on what your macro is set to.  I'd say 'and indicate it is truncated' or something like that.

File: src/cmds/qstat.c [hunk 12]
```diff
@@ -924,10 +962,14 @@ altdsp_statjob(struct batch_status *pstat, struct batch_status *prtheader, int a
 		else
 			timeval = usecput ? eltimecpu : eltimewal;
 
-
 		if (wide) {
+			if (strlen(pstat->name) > SIZEJOBID) {
```
Comment (bhroam) [inline src/cmds/qstat.c]: Why aren't you using your function to handle the truncating?

File: test/tests/functional/pbs_qstat_formats.py [hunk 13]
```diff
@@ -678,3 +678,40 @@ def test_qstat_answ(self):
...
         self.run_namelength_test('-answ')
+
+    def test_qstat_ans(self):
+        """
+        Test if qstat -ans correctly prints with truncation.
+        """
+        self.server.stop()
+        self.assertFalse(self.server.isUp(), 'Failed to stop PBS')
+
+        conf = self.du.parse_pbs_config(self.server.hostname)
```
Comment (bhroam) [inline test/tests/functional/pbs_qstat_formats.py]: You don't need to parse pbs config.  Use self.server.pbs_conf[]

**Comment by vstumpf** [comment]
I'm not sure why I have two travis checks, any ideas how to fix it?

**Comment by mike0042** [comment]
You could try re-signing your commit to restart the build.

**Comment by bhroam** [comment]
LGTM

**Comment by bhroam** [comment]
Please squash and rebase once.  I'll take it from there.

**Pull request merged**